### ISSUE TITLE
Multiple outputs support

### ISSUE THREAD

**Issue by jenbitcoin** [trustwallet/wallet-core#1091]
**Is your feature request related to a problem? Please describe.**
Related to Bip70 support. To be able to set multiple outputs

**Describe the solution you'd like**
Currently, just by setting the **toAddress** the wallet core will already build the lock script for us, and set it as an output.

In this feature, can we be allowed to make our own outputs?
Example, I have this payment output: 

`PaymentOutput(script_hex: "76a914bd80f871f219887af6603711be790200d592829488ac", amount: 546)
`

we'd like to be able to do something like this:

```
var outputs = [TransactionOutput]()
let data = Data(hex: "76a914bd80f871f219887af6603711be790200d592829488ac")
let output = TransactionOutput(value: 546, lockingScript: data)
outputs.append(output)
```

And in creating BitcoinSigningInput, 
we don't need to set the "toAddress" anymore, instead, we will set the value for the 'outputs' -- which is a list of _TransactionOutput_

```
let input = BitcoinSigningInput.with {
    $0.hashType = BitcoinSigHashType.all.rawValue | BitcoinSigHashType.fork.rawValue
    $0.amount = 2000
    $0.byteFee = 1
    $0.outputs = outputs
    $0.changeAddress = "1FQc5LdgGHMHEN9nwkjmz6tWkxhPpxBvBU"
    $0.utxo = [utxo]
    $0.privateKey = [privateKey.data]
}
```

**Comment by stracedude** [comment]
I agree, there is not enough opportunity to make a payment to several addresses at once. 

How do you see the interface for this? ready to help with implementation.

https://blog.coinbase.com/reflections-on-bitcoin-transaction-batching-b13dad12a12

**Comment by jenbitcoin** [comment]
hi @stracedude I read the article, and it's a really good idea to be able to do transaction batching. 👍 
For now, this is how I imagine this enhancement to be like:

```

BitconScript:
- add a function buildWithHexString:
- add a way to be able to append Data

TransactionOutput:
- value (satoshis)
- lockScript data (BitcoinScript)

BitcoinSigningInput:
- add a variable 'outputs' which accepts list of TransactionOutput

```


Example for BIP70:

```
let lockingScript = BitconScript.buildWithHex("76a914bd80f871f219887af6603711be790200d592829488ac")
let output1 = TransactionOutput(value: 1000, script: lockingScript)

let lockingScript2 = BitconScript.buildWithHex("another_hex_string")
let output2 = TransactionOutput(value: 2000, script: lockingScript2)
```

OR, Another example to send to multiple bch addresses:

```
let lockingScript = BitcoinScript.buildForAddress(address: "bitcoincash:qzewrmnw2x5wwdfgj4v5gchf95l8mfpjxgx4ezdsry", coin: .bitcoinCash)
let output1 = TransactionOutput(value: 1000, script: lockingScript)

let lockingScript2 = BitcoinScript.buildForAddress(address: "bitcoincash:qqzxc8pz2epj5ktlfu7fg4nacg6szaqwtcx96mcsll", coin: .bitcoinCash)
let output2 = TransactionOutput(value: 2000, script: lockingScript2)
```


OR, Another example when sending tokens:

```
let script = BitconScript()

do {
try script.append(Data("op_return_data"))
try script.append(Data("token_id"))
try script.append(Data("token_amount"))
}

let output1 = TransactionOutput(value: 0, script: script) // OP_RETURN

let lockingScript = BitcoinScript.buildForAddress(address: "simpleledger:qzewrmnw2x5wwdfgj4v5gchf95l8mfpjxgx4ezdsry", coin: .bitcoinCash)
let output2 = TransactionOutput(value: 546, script: lockingScript)
```


Lastly, to be able to use these outputs (in my example, there are 2 outputs), we can set this to BitcoinSigningInput

```
let input = BitcoinSigningInput.with {
    $0.hashType = BitcoinSigHashType.all.rawValue | BitcoinSigHashType.fork.rawValue
    $0.amount = 3000
    $0.byteFee = 1
    $0.outputs = [output1, output2]
    $0.changeAddress = "1FQc5LdgGHMHEN9nwkjmz6tWkxhPpxBvBU"
    $0.utxo = [utxo]
    $0.privateKey = [privateKey.data]
}
```

I hope it makes sense :)

**Comment by catenocrypt** [comment]
@jenbitcoin , could you elaborate how is this related to BIP70?

**Comment by jenbitcoin** [comment]
Hi @catenocrypt 

Our app allows paying to merchants. So, if I'm a merchant, I ask the customer to pay me specific amount, and create an invoice for that customer. Here's a sample invoice: https://pay.bitcoin.com/i/7wdChtjM6Xu1EhVaJW44Dn (This invoice will expire in 15mins)

The user who uses our app can scan this QR, the app determines that this is a merchant address, so the app asks for the payment information from our backend.

The backend will return list of PaymentOutputs, that look like this:
`PaymentOutput(script_hex: "76a914bd80f871f219887af6603711be790200d592829488ac", amount: 546)`

Then we process the payment outputs.
This article might also help: https://bitcoinj.github.io/payment-protocol

**Comment by catenocrypt** [comment]
Exchanges have been doing ‘transaction batching’ for long, with multi-output transactions.  This coinbase article explains it nicely, with example and backup data.
TrustWallet, being primarily for wallets, does not currently support multi-output transaction building.  It could be added, as suggested, but:
- it should not make the common 1/2 output case more complex to use in any way
- lock script building for multi-output case should be also supported
- arbitrary script building may not be justified at this point (weighing development complexity vs. usage)
- multiple method calls is a nice way for handling variable number of outputs. However, this is probably captured in a single underlying protobuf message, with an array of outputs.

**Comment by catenocrypt** [comment]
Here is the relevant code that builds TX with 2 or 1 outputs.  More outputs are not supported in the protobuf message, and in the 'public' C (and Switft/Kotlin) interfaces, but multiple outputs are supported in the lower layers.

Functionality could be exposed by generalizing the exisiting protobuf, or creating a new extended version.

https://github.com/trustwallet/wallet-core/blob/master/src/Bitcoin/TransactionBuilder.h#L24

```C++
    /// Builds a transaction by selecting UTXOs and calculating fees.
    template <typename Transaction>
    static Transaction build(const TransactionPlan& plan, const std::string& toAddress,
                             const std::string& changeAddress, enum TWCoinType coin) {
        auto lockingScriptTo = Script::lockScriptForAddress(toAddress, coin);
        if (lockingScriptTo.empty()) {
            return {};
        }

        Transaction tx;
        tx.outputs.push_back(TransactionOutput(plan.amount, lockingScriptTo));

        if (plan.change > 0) {
            auto lockingScriptChange = Script::lockScriptForAddress(changeAddress, coin);
            tx.outputs.push_back(TransactionOutput(plan.change, lockingScriptChange));
        }

        const auto emptyScript = Script();
        for (auto& utxo : plan.utxos) {
            tx.inputs.emplace_back(utxo.out_point(), emptyScript, utxo.out_point().sequence());
        }

        return tx;
    }
```

**Comment by jenbitcoin** [comment]
@catenocrypt 

> arbitrary script building may not be justified at this point (weighing development complexity vs. usage)

I think we'll be fine if we omit script building. Our team will find a way to fix this on our end :)

**Comment by catenocrypt** [comment]
Multi-output support has been added, in PR #1103 .
@jenbitcoin , please have a look at the Swift test:

https://github.com/trustwallet/wallet-core/blob/ar/multi-output/swift/Tests/Blockchains/BitcoinTests.swift#L125

(it is techincally a litecoin TX, but code is the same)

**Comment by jenbitcoin** [comment]
thanks @catenocrypt 🥳 - will check this out today.

**Comment by jenbitcoin** [comment]
@catenocrypt  i tried downloading this branch to my project (via Pods) so i can test it on our end, but I'm getting this error instead:

```
Pre-downloading: `TrustWalletCore` from `https://github.com/trustwallet/wallet-core.git`, branch `ar/multi-output`
[!] Failed to download 'TrustWalletCore': [!] /bin/bash -c 
set -e
tools/install-dependencies && tools/generate-files

PREFIX: /Users/jenn/Library/Caches/CocoaPods/Pods/External/TrustWalletCore/3354d06c37a9a2d703ad708b97b472d7/build/local
tools/install-dependencies: line 29: cmake: command not found
```

**Comment by stracedude** [comment]
@jenbitcoin You need to install cmake using your package manager. (brew install cmake).
https://developer.trustwallet.com/wallet-core/developing-the-library/building

**Comment by stracedude** [comment]
@catenocrypt It is very sad that the changes were closed. I will use your code in my project and try to come up with a better API for adding this functionality.

**Comment by catenocrypt** [comment]
Please note that multi-output TX is possible to put together if using the C++ parts directly, not through protobuf messages. This is possible form C++ or GoLang integration, but not Swift or Kotlin bindings.
For now, a sample test is being added (#1106 )
